### PR TITLE
1.Added method for zcb parsing 2. Refactored ResponseSegment 3. fixed…

### PR DESCRIPTION
**Pull request by AnumehaSrivastava05**
… validation error 4. Created datetime method 5. created method to populate optional field

### CODE REVIEW

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/message/ResponseSegment.java [hunk 1]
```diff
@@ -1,9 +1,99 @@
...
-public interface ResponseSegment {
-	
-	String constructMSH(HL7Message messageObj);
-	
-	String constructResponse(HL7Message messageObj, ErrorMessage error);
+import java.util.Optional;
 
+import ca.bc.gov.hlth.hnsecure.parsing.Util;
+
+public abstract class ResponseSegment {
```
[the remark this answers is not in the document]
Comment (AnumehaSrivastava05) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/message/ResponseSegment.java:7]: constructResponse() implementation will change as per messageType

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/message/PharmanetErrorResponse.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		sb.append(TRANSACTION_CODE);
+		sb.append(getFieldseperator());
+		sb.append(getFieldseperator());
+		sb.append(getFieldseperator());
+		sb.append("\n");
+		return sb.toString();
+	}
+
+
+	private String buildZCB() {
```
[the remark this answers is not in the document]
Comment (AnumehaSrivastava05) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/message/PharmanetErrorResponse.java:37]: This method might be modified in order to build zcb segment for other errors(eg: encryption protocol error or facility id mismatch). Legacy code populates field too.(Need to verify)

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/messagevalidation/V2PayloadValidator.java [hunk 3]
```diff
@@ -51,8 +53,119 @@ public static void validate(Exchange exchange, String v2Message) throws Validati
...
 
 		// Validate v2Message format
+		validateMessageFormat(exchange, v2Message, messageObj);
+
+		boolean isPharmanetMode = isPharmanet(messageObj);
+
+		validateSendingFacility(exchange, messageObj, accessToken, isPharmanetMode);
+
+		validateReceivingApp(exchange, messageObj);
+
```
Comment (pankaj-cgi) [inline hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/messagevalidation/V2PayloadValidator.java:63]: Should add Junits for new messages

File: hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/message/PharmanetErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	private static final String TRANSACTION_CODE = "50";
+
+	@Override
+	public String constructResponse(HL7Message messageObj, ErrorMessage error) {
+		
+		return constructMSH(messageObj)+buildZCA()+buildZCB()+buildZZZ(error);
+	}
+	
+	private String buildZCA() {
+		StringBuilder sb = new StringBuilder(ZCA_IDENTIFIER );
```
[the remark this answers is not in the document]
Comment (pankaj-cgi) [reply hnsecure/src/main/java/ca/bc/gov/hlth/hnsecure/message/PharmanetErrorResponse.java:38]: Added comment

**Pull request merged**